src/critics/local_goal_critic.cpp

```diff
@@ -0,0 +1,113 @@
+// Copyright 2022 @artofnothingness Alexey Budyakov, Samsung Research
+#include "mppic/critics/local_goal_critic.hpp"
+
+#include <xtensor/xmath.hpp>
+#include <xtensor/xsort.hpp>
+
+namespace mppi::critics
+{
+
+void LocalGoalCritic::initialize()
+{
+  auto getParam = parameters_handler_->getParamGetter(name_);
+
+  getParam(consider_angle_, "consider_angle", true);
+  getParam(angle_offset_, "angle_offset", 6);
+  getParam(angle_cost_power_, "angle_cost_power", 1);
+  getParam(angle_cost_weight_, "angle_cost_weight", 3.0);
+
+  getParam(consider_distance_, "consider_distance", true);
+  getParam(distance_goal_count_, "distance_goal_count", 2);
+  getParam(distance_offset_, "distance_offset", 6);
+  getParam(distance_cost_power_, "distance_cost_power", 2);
+  getParam(distance_cost_weight_, "distance_cost_weight", 3.0);
+
+  getParam(stop_usage_path_reached_ratio_, "stop_usage_path_reached_ratio", 0.35);
+}
+
+void LocalGoalCritic::score(models::CriticFunctionData & data)
+{
+  if(!enabled_) {
+    return;
+  }
+
+  auto path_points = xt::view(data.path, xt::all(), xt::range(0, 2));
+
+  // get furthest point among trajectories nearest points to path
+  size_t furthest_reached_path_point_idx;
+  if (data.furthest_reached_path_point) {
+    furthest_reached_path_point_idx = *data.furthest_reached_path_point;
+  } else {
+    furthest_reached_path_point_idx = findPathFurthestPoint(data);
+  }
+
+  // calculate range of points to consider as local goals
+  auto lower_distance_offset = std::min(furthest_reached_path_point_idx + distance_offset_, path_points.shape(0) - 1);
+  auto upper_distance_offset = std::min(lower_distance_offset + distance_goal_count_, path_points.shape(0) - 1);
+
+  // calculate angle point as local goal
+  auto angle_offset = std::min(
+    furthest_reached_path_point_idx + angle_offset_, path_points.shape(
+      0) - 1);
+
+  // portion of the path which if reached by trajectories then critic returns
+  size_t threshold_idx = static_cast<size_t>(
+    static_cast<double>(data.path.shape(0) - 1) * stop_usage_path_reached_ratio_);
+
+  if (lower_distance_offset < threshold_idx) {
+    return;
+  }
+
+  if (consider_distance_ ) {
+    auto path_interval = xt::view(path_points, xt::range(lower_distance_offset, upper_distance_offset), xt::all());
+    auto trajectories_last_points = xt::view(data.trajectories, xt::all(), -1, xt::newaxis(), xt::range(0, 2));
+    auto distance_to_furthest = xt::norm_l2(trajectories_last_points - path_interval, {2});
+    auto mean_distance_to_furthest = xt::mean(distance_to_furthest, {1});
+    data.costs += xt::pow(distance_cost_weight_ * mean_distance_to_furthest, distance_cost_power_);
+  }
+
+  // GoalAngleCritic for local goal
+  if (consider_angle_) {
```

Comment:
Here's my suggestion: 

It seems like this is just the same thing as `PathAngleCritic`, this is really for the same thing as that - e.g. the role of this bit of code is to align the path angle.  

Have you tried adding the `angle_offset` to `PathAngleCritic`? That would be the same thing as this below, and would eliminate this term from this critic entirely if it can be replaced. That would be the most modular. 

Right now, the default configuration doesn't use `PathAngleCritic` which makes me think we should just add the offset calculation to that critic and include it in the configuration instead. 

Presto! That would decouple things really nicely.

Reply:
Actually i do realize now that PathAngleCritic really described much better with AngleToGoalCritic, coz it's basically angle to the goal (last path point) now. 

So here it is the same connection as with GoalCritic and current critic distance part. 

Goal Critic and PathAngleCritic (AngleToGoalCritic) penalize according to the last point. 

This one (PathFollowCritic) penalize for the path point that is furthest from current trajectories. So technically, they did the same, but conditions are very different, and also distance part of this critic uses mean of some point count. 

We could replace PathAngleCritic with current critic angle part though. I don't think we need AngleToGoalCritic (current PathAngleCritic)

Reply:
>  coz it's basically angle to the goal (last path point) now.

Last thresholded path from the path handler, so still a path point ('local goal')

I'm just talking about specifically the part of this critic that is `consider_angle_`. That is the same as PathAngleCritic but with a different way of finding the point. I think you should remove the `consider_angle_` here entirely and update this logic in PathAngleCritic -- then use that critic in the default bringup. 

That would make this critic then only `consider_distance_`'s stuff, which then could be renamed more usefully to what that role is  (see below, I think what the `consider_distance_` stuff does could relate to #50)

Reply:
Last thresholded path from the path handler, so still a path point ('local goal')

Yeah that's true xD 
But GoalCritic actually has the same prefix Goal, although it uses the same last point of thresholded path 
PathAngleCritic now uses the same point as GoalCritic.

> I'm just talking about specifically the part of this critic that is consider_angle_. That is the same as PathAngleCritic but with a different way of finding the point. I think you should remove the consider_angle_ here entirely and update this logic in PathAngleCritic -- then use that critic in the default bringup. Even just looking at PathAngleCritic versus the stuff in consider_angle_, I think consider_angle_ would do a better job at the task path angle which PathAngleCritic should be doing.

Yeah, i got the idea, and i like it.
I just wanted to mention that current PathAngleCritic uses the same point as GoalCritic, and probably must be name AngleToGoal by what it does. Anyway i agree here

Reply:
> PathAngleCritic now uses the same point as GoalCritic

Sounds like a perfect thing than to remove `consider_angle_` from here than and replace it `PathAngleCritic` with it. 

>  I just wanted to mention that current PathAngleCritic uses the same point as GoalCritic, and probably better be named named AngleToGoal by what it does.

I'd keep the names -- there is also a [goal angle critic](https://github.com/artofnothingness/mppic/blob/develop/src/critics/goal_angle_critic.cpp). I think PathAngle / PathAlign (other PR) and Goal / Goal Angle are perfect and descriptive names. I wouldn't change anything about the `goal` ones.

Reply:
separate them